### PR TITLE
Improve handling of multiple short options by supporting short syntax

### LINKED ISSUE #1: CLI: improve handling of multiple short options
https://github.com/hhvm/hhast/blob/master/src/__Private/CLIBase.php

It's normal to be able to combine multiple short options into a single one - for example, `-a -b` should be equivalent to `-ab`.

Complications:
 - this shouldn't be true for short options which require an argument - for example, if b needs an option, `-a -b foo` is required  - not `-abc foo` or `-ab foo`
 - this will require banning multi-character short options - you can do this in the CLIOption base class with an invariant
 - that means that `-vv` will need to be removed from `bin/hhast-lint`; this can be done by changing `-v` to increment `$verbosity`, instead of setting it to 1

### PR DESCRIPTION
**Pull request by YuryBandarchuk16**
Fix #1

### CODE REVIEW

File: src/__Private/CLIBase.php [hunk 1]
```diff
@@ -206,6 +208,48 @@ public function __construct(
...
+      if (Str\contains($option, '=')) {
+        $parts = \explode('=', $option);
+        $opt = $parts[0];
+        $value = \implode('=', Vec\drop($parts, 1));
+      }
+      $result = dict[
+        $option => $value,
+      ];
+    } else {
+      $result = Dict\pull(
```
Comment (kmeht) [inline src/__Private/CLIBase.php]:
```
$result = Dict\fill_keys($options, null);
```

File: src/__Private/CLIBase.php [hunk 2]
```diff
@@ -206,6 +208,48 @@ public function __construct(
...
+        break;
+
+      default:
+        break;
+    }
+
+    $is_single_option = C\count($options) === 1;
+    if ($is_single_option) {
+      $value = null;
+      $option = C\firstx($options);
```
Comment (kmeht) [inline src/__Private/CLIBase.php]: Optional: `C\onlyx`.

File: src/__Private/CLIBase.php [hunk 3]
```diff
@@ -206,6 +208,48 @@ public function __construct(
...
+    switch ($type) {
+      case CLIOptions\CLIOptionType::LONG:
+        $options[] = $arg;
+        break;
+
+      case CLIOptions\CLIOptionType::SHORT:
+        $arg_length = Str\length($arg);
+        for ($index = 0; $index < $arg_length; $index++) {
+          $options[] = $arg[$index];
+        }
```
Comment (kmeht) [inline src/__Private/CLIBase.php]:
```
$options = Vec\concat($options, Str\chunk($arg));
```

File: src/__Private/CLIBase.php [hunk 4]
```diff
@@ -206,6 +208,48 @@ public function __construct(
...
+      default:
+        break;
+    }
+
+    $is_single_option = C\count($options) === 1;
+    if ($is_single_option) {
+      $value = null;
+      $option = C\firstx($options);
+      if (Str\contains($option, '=')) {
+        $parts = \explode('=', $option);
```
Comment (kmeht) [inline src/__Private/CLIBase.php]: `Str\split($option, '=')`

File: src/__Private/CLIBase.php [hunk 5]
```diff
@@ -206,6 +208,48 @@ public function __construct(
...
+        $opt = $parts[0];
+        $value = \implode('=', Vec\drop($parts, 1));
```
Comment (kmeht) [inline src/__Private/CLIBase.php]: `Str\join(Vec\drop($parts, 1), '=')`

File: src/__Private/CLIBase.php [hunk 6]
```diff
@@ -206,6 +208,48 @@ public function __construct(
...
+        break;
+
+      case CLIOptions\CLIOptionType::SHORT:
+        $arg_length = Str\length($arg);
+        for ($index = 0; $index < $arg_length; $index++) {
+          $options[] = $arg[$index];
+        }
+        break;
+
+      default:
```
Comment (kmeht) [inline src/__Private/CLIBase.php]: Can remove this, Hack can do exhaustiveness checking on enums.

Comment (YuryBandarchuk16) [reply src/__Private/CLIBase.php]: Actually, I omitted one enum value so could not remove default branch

**Comment by fredemmott** [comment]
I'm going to hold off on merging this for a few days; I'm doing some work that would make the CLI library testable, which I think is needed before merging this.

**Comment by fredemmott** [comment]
This is going to be resubmitted against hh-clilib

**Pull request closed without merging**